doc/src/manual/types.md

```diff
@@ -1108,6 +1108,47 @@ julia> NoFieldsParam{Int}() === NoFieldsParam{Int}()
 true
 ```
 
+## Types of functions and closures (anonymous functions)
+
+Each function has its own type, which is a subtype of `Function`.
+
+```jldoctest
+julia> foo41(x) = x + 1
+foo (generic function with 1 method)
+
+julia> typeof(foo41)
+typeof(foo41) (singleton type of function foo41, subtype of Function)
+```
+
+Note how `typeof(foo41)` prints as itself. This is merely a convention for printing, as it is a first-class object that can be used like any other value:
+
+```jldoctest
+julia> T = typeof(foo41)
+typeof(foo41) (singleton type of function foo41, subtype of Function)
+
+julia> T <: Function
+true
+```
+
+Types of functions are singletons. When necessary, you can compare them with [`===`](@ref).
+
+[Anonymous functions](@ref man-anonymous-functions) also have their own type, which is printed with names that include `#<number>`. Numbers for different function types are distinct, but not guaranteed to be the same across sessions.
+
+```jldoctest; filter = r"[0-9\.]+"
+julia> typeof(x -> x + 1)
+var"#9#10"
+```
+
+Types of anonymous function are not necessarily singletons, for example in cases where a variable is captured.
+
+```jldoctest
+julia> addy(y) = x -> x + y
+addy (generic function with 1 method)
+
+julia> Base.issingletontype(addy(1))
+false
```

Comment:
```suggestion
false

julia> addy(1).y
1
```

Reply:
Isn't closures having fields an implementation detail, something we would not necessarily want to expose in the user manual?

Reply:
I wouldn't call that an implementation detail. It is a defining feature of closures.

Reply:
It doesn't sound like a defining feature of closures, after all they were implemented differently in the past, and the field names could have been made unavailable through `getproperty` even with the current implementation. But I don't know whether that's now specified somewhere in the documentation.

Reply:
It's in the [dev docs](https://docs.julialang.org/en/v1/devdocs/functions/#Closures). 

Incidentally, is it OK to link devdocs in the user manual?

Reply:
> It doesn't sound like a defining feature of closures, after all they were implemented differently in the past, and the field names could have been made unavailable through getproperty even with the current implementation.

I actually didn't know that, thanks! I suspect that changing this now would be quite breaking, so we can probably document this somewhere, but we can leave it out of this PR for now.

Reply:
> Incidentally, is it OK to link devdocs in the user manual?

I didn't notice that at first. We should probably have something on closures n the regular docs as well, but for now, maybe we can just have the link point to `anonymous functions` instead, since in most cases they are actually closures as well.

Reply:
> It's in the dev docs.

Thanks. As the "dev docs" document Julia internals, I wouldn't link to the them from the user manual. To show the point here, you might insert in the sessions `addy(1) === addy(2) # false`, and/or maybe show the output of `addy(1)` (i.e. `var"#3#4"{Int64}(1)`, which hints at the implementation detail).

Reply:
Thanks, I added that example, and now only refer to closures in passing.